### PR TITLE
[wip] Optimization of smoothening of loss function

### PR DESCRIPTION
**Pull request by purvaten**

### CODE REVIEW

File: patch_recurrence/config/arguments.py~ [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import argparse
```
Comment (martiansideofthemoon) [inline patch_recurrence/config/arguments.py~]: Remove file

File: patch_recurrence/config/constants.yml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 K_NEAREST: 9
 PATCH_SIZE: 7
-PAIR_THRESHOLD: 0.8
+PAIR_THRESHOLD: 0.7
```
Comment (martiansideofthemoon) [inline patch_recurrence/config/constants.yml:3]: Any reason for this?

File: patch_recurrence/dehaze.py [hunk 3]
```diff
@@ -57,8 +62,11 @@ def main():
     else:
         patches, pairs = None, None
     if patches is None and pairs is None:
-        logger.info("Extracting all patches ...")
-        patches = steps.generate_patches(scaled_imgs, constants, True)
+        logger.info("Extracting alternate patches ...")
+        patches = steps.generate_patches(scaled_imgs, constants, False)
+
+        print("\nNumber of patches extracted per scaled image")
+        print(len(patches[0]), len(patches[1]), len(patches[2]), len(patches[3]), len(patches[4]), len(patches[5]))
```
Comment (martiansideofthemoon) [inline patch_recurrence/dehaze.py]: `print(map(len, patches))`

File: patch_recurrence/dehaze.py [hunk 4]
```diff
@@ -77,13 +88,34 @@ def main():
...
     logger.info("Estimating global airlight ...")
     airlight = steps.estimate_airlight(pairs)
 
     logger.info("Estimated airlight is ...%s", str(airlight))
 
+    # T-map estimation code begins
+
+    del pairs
+    del patches
+    gc.collect()
```
Comment (martiansideofthemoon) [inline patch_recurrence/dehaze.py]: Does this help?

File: patch_recurrence/patch.py [hunk 5]
```diff
@@ -45,12 +45,14 @@ def __init__(self, first, second):
         raw2 = np.reshape(second.raw_patch, [-1, 3])
         # Equation (11) in Bahat et al.
         self.airlight = np.zeros(3)
+        np.seterr(all='raise')
```
Comment (martiansideofthemoon) [inline patch_recurrence/patch.py]: What is this for?

File: patch_recurrence/steps.py [hunk 6]
```diff
@@ -22,20 +44,19 @@ def scale(img, scales):
 
 def generate_patches(scaled_imgs, constants, all_patches):
     patch_size = constants.PATCH_SIZE
-    std_dev_threshold = constants.STD_DEV_THRESHOLD
+    step = 1 if all_patches else 2
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: Maybe move `all_patches` to configuration?

File: patch_recurrence/steps.py [hunk 7]
```diff
@@ -22,20 +44,19 @@ def scale(img, scales):
...
+            for j in range(0, sc.shape[1] - patch_size, step):
                 raw_patch = sc[i:i + patch_size, j:j + patch_size, :]
                 patch = Patch(
                     raw_patch=raw_patch,
                     patch_size=patch_size,
                 )
-                if all_patches or patch.std_dev > std_dev_threshold:
-                    patch.store(sc, [i, j])
-                    img_patches.append(patch)
+                patch.store(sc, [i, j])
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py:49]: not checking `std_dev_threshold`?

File: patch_recurrence/steps.py [hunk 8]
```diff
@@ -50,8 +71,12 @@ def smoothen(scaled_imgs, patches, constants):
         img = scaled_imgs[k]
         patch = patches[k]
 
-        length_sd_array = img.shape[0] - patch_size
-        width_sd_array = img.shape[1] - patch_size
+        # For half patches
+        length_sd_array = width_sd_array = 0
+        for i in range(0, img.shape[0] - patch_size, 2):
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: I didn't understand this part. Is `2` hard-coded?

File: patch_recurrence/steps.py [hunk 9]
```diff
@@ -60,25 +85,31 @@ def smoothen(scaled_imgs, patches, constants):
...
 
-    std_database = np.zeros((scaled_imgs, width))
     for k in range(scaled_imgs):
-        for index, patch in enumerate(patches[k]):
-            std_database[k][index] = patch.std_dev
+        std_database = []
+        patch_database = []
+        for patch in patches[k]:
+            std_database.append(patch.std_dev)
+            patch_database.append(0)
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: you can just say `patch_database = [0] * len(patches[k])`

File: patch_recurrence/steps.py [hunk 10]
```diff
@@ -60,25 +85,31 @@ def smoothen(scaled_imgs, patches, constants):
...
 
-    min_std = np.amin(std_database)
-    max_std = np.amax(std_database)
+        index = np.argsort(std_database)
+        interval = len(std_database) / num_buckets
 
-    norm_std = np.floor((std_database - min_std) * (num_buckets - 1) / (max_std - min_std))
-    for k in range(scaled_imgs):
-        map(lambda (i, x): setattr(x, 'bucket', norm_std[k][i]), enumerate(patches[k]))
+        for i in range(num_buckets):
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: You might miss a few patches in this loop. I think this can be done in a cleaner way, but if it's not time consuming it's fine

File: patch_recurrence/steps.py [hunk 11]
```diff
@@ -159,3 +202,101 @@ def estimate_airlight(pairs):
         numerator += pair.weight * pair.airlight
         denominator += pair.weight
     return (numerator / denominator)
+
+
+# Tmap estimation functions
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: Move PyTorch to different file

File: patch_recurrence/steps.py [hunk 12]
```diff
@@ -88,12 +119,24 @@ def generate_pairs(imgs, patches, constants):
     index_database = []
     length_database = []
     for k in range(scaled_imgs):
+        a = [patch.norm_patch for patch in patches[k] if 7 <= patch.bucket <= 9]
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: this part is hard to read, what exactly are you trying to do here? We should clean up the parts with `x` and `t`

File: start_patch_recurrence.sh [hunk 13]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 
-python patch_recurrence/dehaze.py "$@"
+python2 patch_recurrence/dehaze.py "$@"
```
Comment (martiansideofthemoon) [inline start_patch_recurrence.sh]: is this necessary?

File: patch_recurrence/steps.py [hunk 14]
```diff
@@ -159,3 +202,101 @@ def estimate_airlight(pairs):
...
+    return t
+
+
+def estimate_tmap(img, patches, airlight, constants):
+    """Estimates t-map and returns dehazed output image after 20 iterations
+    """
+    patch_size = constants.PATCH_SIZE
+    tlb = np.empty([len(patches)])
+    for index, patch in enumerate(patches):
+        raw = np.reshape(patch.raw_patch, [-1, 3])
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: Add some comments on equation number / what u are doing

File: patch_recurrence/config/constants.yml (hunk 2 above)
Comment (purvaten) [reply patch_recurrence/config/constants.yml:3]: I thought the threshold was too strict because initially I wasn't getting too many pairs. I will see if this has changed now and if so change it back to 0.8

File: patch_recurrence/dehaze.py (hunk 4 above)
Comment (purvaten) [reply patch_recurrence/dehaze.py]: Yes it definitely does. But I still think that instead of re-obtaining all patches, I should try to get them all at once in the first step itself. I should then make use of alternate patches for the initial steps and later make use of all of them for t-map estimation. I will try to remove this unnecessary repetition.

File: patch_recurrence/steps.py (hunk 6 above)
Comment (purvaten) [reply patch_recurrence/steps.py]: Nope. The reason I did this is because I am considering alternate patches of all scaled images initially (all_patches = 0). Later for the t-map estimation, I require all patches of just the original image (all_patches = 1). I will try to see if there is a better way to do this.

File: patch_recurrence/steps.py (hunk 10 above)
Comment (purvaten) [reply patch_recurrence/steps.py]: Takes less than a second. Why will I miss a few patches? Yes, I agree it can be done in a cleaner way. Will rewrite it.

File: patch_recurrence/steps.py (hunk 7 above)
Comment (martiansideofthemoon) [reply patch_recurrence/steps.py:49]: ?

File: patch_recurrence/steps.py (hunk 8 above)
Comment (martiansideofthemoon) [reply patch_recurrence/steps.py]: ?

File: patch_recurrence/steps.py (hunk 7 above)
Comment (purvaten) [reply patch_recurrence/steps.py:49]: I thought we had already removed this. Why is there even a srd_dev_threshold in the first place? Aren't we considering all patches?

File: patch_recurrence/steps.py (hunk 8 above)
Comment (purvaten) [reply patch_recurrence/steps.py]: Yeah it was for alternate patches - I am changing it to a variable like all_patches in generate_patches.

File: patch_recurrence/steps.py (hunk 7 above)
Comment (purvaten) [reply patch_recurrence/steps.py:49]: Actually I get your point. Were we doing it to ensure the patches are at least a little informative, i.e. not fully white?

File: patch_recurrence/tmap_steps.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
+"""Transmission map estimation functions."""
+import numpy as np
+
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py:3]: **nit** - remove unnecessary blank lines

File: patch_recurrence/tmap_steps.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+        super(Net, self).__init__()
+        self.img = img
+        self.patches = patches
+        self.airlight = airlight
+        self.constants = constants
+
+    def sigmoid(self, y, length):
+        sig = torch.Tensor(length).type(dtype).zero_()
+        for i in range(length):
+            l2_norm = math.sqrt(sum(map(lambda x: x * x, y[i])))
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: why don't you move this computation to `torch`?

File: patch_recurrence/tmap_steps.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        l2_norm = self.get_norm(w)
+        s = l2_norm * sig
+        ret = torch.sum(s)
+        return ret
+
+
+def estimate_tmap(img, patches, airlight, constants):
+    """Estimates t-map and returns dehazed output image after 20 iterations
+    """
+    net = Net(img, patches, airlight, constants)
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: **nit** - is there any need to pass all those parameters?

File: patch_recurrence/tmap_steps.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    sig = sig.view(h - patch_size, w - patch_size)
+
+    # Actual Optimization
+    t_height, t_width = len(tlb), len(tlb[0])
+    tlb = torch.Tensor(np.reshape(tlb, [t_height, t_width])).type(dtype)
+    for i in range(t_height):
+        for j in range(t_width):
+            if tlb[i][j] <= 0:
+                tlb[i][j] = 10 ** -7
+    weight = Variable(tlb, requires_grad=True)
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: you should define this in the `__init__` function of your `nn.Module` subclass

File: patch_recurrence/tmap_steps.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    sig = Variable(sig, requires_grad=False)
+    optimizer = torch.optim.SGD([weight], lr=0.001)
+    for i in range(100):
+        optimizer.zero_grad()
+        loss = net(weight, sig)
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: you should not be passing `weight` and `sig` here. They should be defined inside the class's `__init__` functions. Also you can avoid making `sig` a `Variable` with some effort

File: patch_recurrence/tmap_steps.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+        logger.info("Loss is %f", loss)
+        loss.backward()
+        optimizer.step()
+        tmap = weight.data
+        # Recalculate sig and weight based on tmap
+        weight = Variable(tmap, requires_grad=True)
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: why do you need to do this? `weight` already stores `tmap`. In TensorFlow, doing this would make the code extremely slow.

File: patch_recurrence/tmap_steps.py [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+        tmap = tmap.cpu().numpy()
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: why `.cpu()`?

File: patch_recurrence/tmap_steps.py [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+        tmap = np.reshape(tmap, [h - patch_size, w - patch_size, 1])
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: can you move all `tmap` ---> `l_img` steps to a different function? Ideally, for subsequent `sig` calculation, you would want to make this whole thing end-to-end

File: patch_recurrence/patch.py [hunk 23]
```diff
@@ -21,7 +21,7 @@ def __init__(self, raw_patch, patch_size):
 
     def store(self, img, location):
         # If the patch passes std_dev test, the following vector used in KNN
-        self.norm_patch = self.mean_free_patch / self.std_dev
+        self.norm_patch = 0 if self.std_dev == 0 else self.mean_free_patch / self.std_dev
```
Comment (martiansideofthemoon) [inline patch_recurrence/patch.py]: let's use `np.close()` instead of `==`

File: patch_recurrence/steps.py (hunk 8 above)
Comment (martiansideofthemoon) [reply patch_recurrence/steps.py]: can you change this to `step`? Ideally we want to move the `step` size to the configuration file. Also you don't need this loop at all. `length_sd_array` will be `(img.shape[0] - patch_size) / 2`

File: patch_recurrence/steps.py [hunk 24]
```diff
@@ -50,8 +61,12 @@ def smoothen(scaled_imgs, patches, constants):
         img = scaled_imgs[k]
         patch = patches[k]
 
-        length_sd_array = img.shape[0] - patch_size
-        width_sd_array = img.shape[1] - patch_size
+        # For half patches
+        length_sd_array = width_sd_array = 0
+        for i in range(0, img.shape[0] - patch_size, 2):
+            length_sd_array += 1
+        for i in range(0, img.shape[1] - patch_size, 2):
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: `step` here as well. Again, we don't need the loop

File: patch_recurrence/tmap_steps.py [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+        conv1 = F.conv2d(x, diff1, padding=(0, 1))
+        conv2 = F.conv2d(x, diff2, padding=(1, 0))
+        return conv1, conv2
+
+    def get_norm(self, x):
+        """Return l2-norm squared for grad(log(t))."""
+        height, width = x.size(0), x.size(1)
+        log = torch.log(torch.clamp(x, min=0.0000001, max=1))
+        log = log.view(1, 1, height, width)
+        conv1, conv2 = self.gradient(log)
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py]: let's call this function `spatial_gradient`

File: patch_recurrence/tmap_steps.py [hunk 26]
```diff
@@ -0,0 +1,163 @@
+"""Transmission map estimation functions."""
```
Comment (martiansideofthemoon) [inline patch_recurrence/tmap_steps.py:1]: This file is well-written overall but needs more comments. Let's add more comments! Since I'm not critically reviewing this code now, this can be done after the merge

File: patch_recurrence/steps.py [hunk 27]
```diff
@@ -135,7 +175,7 @@ def filter_pairs(patches, pairs, constants):
         )
         correlation = 1 - distance
 
-        if correlation >= pair_threshold:
+        if correlation >= pair_threshold or all_pairs:
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py:205]: Since the only reason to put `all_pairs` here is to create `Pair` objects, maybe we could move `Pair` object creation to a separate function after this

File: patch_recurrence/steps.py (hunk 10 above)
Comment (martiansideofthemoon) [reply patch_recurrence/steps.py]: You will miss patches if `len(std_database)` is not perfectly divisible by `num_buckets`, since this loop iterates over `num_buckets * interval` elements

File: patch_recurrence/steps.py [hunk 28]
```diff
@@ -88,15 +107,36 @@ def generate_pairs(imgs, patches, constants):
     index_database = []
     length_database = []
     for k in range(scaled_imgs):
+        if raw:
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py]: Add a comment here, and in general in all the confusing parts of the code

**Review by martiansideofthemoon — COMMENTED** [review]
Overall code is fine. I would like to see more comments in the `steps.py` file. We can try to make the code a bit more modular by making a `PatchDatabase` class and a `PairDatabase` class and dividing the functions in `steps.py` between them

File: patch_recurrence/steps.py (hunk 24 above)
Comment (purvaten) [reply patch_recurrence/steps.py]: I thought so too about the loop, but no, it doesn't work directly. It depends on whether the img.shape[0] and img.shape[1] are even or odd. It changes for every image. Let me know if you can think of a better way.

Comment (purvaten) [reply patch_recurrence/steps.py]: I can rewrite it by removing the for loop and putting a condition for odd even stuff

Comment (martiansideofthemoon) [reply patch_recurrence/steps.py]: No it doesn't, just use integer division (3 / 2 = 1, 2 / 2 = 1)

Comment (purvaten) [reply patch_recurrence/steps.py]: Yup sorry my bad

Comment (purvaten) [reply patch_recurrence/steps.py]: The reason I didn't include step in constants and instead used a variable all_patches was because for the initial step I am extracting alternate patches. But later I am extracting all patches for the raw PatchMatch step. I understand of course that is resulting in unnecessary repetition. Will try to fix that later.

File: patch_recurrence/steps.py (hunk 27 above)
Comment (purvaten) [reply patch_recurrence/steps.py:205]: Yeah I think so too. Should I do it now or after the merge?

File: patch_recurrence/steps.py [hunk 29]
```diff
@@ -21,65 +33,72 @@ def scale(img, scales):
...
     patch_size = constants.PATCH_SIZE
 
     for k in range(len(patches)):
         img = scaled_imgs[k]
         patch = patches[k]
 
-        length_sd_array = img.shape[0] - patch_size
-        width_sd_array = img.shape[1] - patch_size
+        # We assume that alternate patches have been extracted in the initial step
+        length_sd_array = int(round((img.shape[0] - patch_size) / 2))
```
Comment (martiansideofthemoon) [inline patch_recurrence/steps.py:64]: this should be `step` right?

File: patch_recurrence/tmap_steps.py (hunk 25 above)
Comment (martiansideofthemoon) [reply patch_recurrence/tmap_steps.py]: I was referring to other gradient function :smile:
